File: src/main/scala/kim/dencka/easy/linked_list/ReverseLinkedList.scala
package kim.dencka.easy.linked_list

import kim.dencka.payloads.LINKED_LIST_STRUCTURE._

object ReverseLinkedList extends App {

  def reverseList(head: ListNode): ListNode = {
    @annotation.tailrec
    def run(ln: ListNode, acc: ListNode): ListNode = ln match {
      case null => acc
      case _ => run(ln.next, new ListNode(ln.x, acc))
    }
    run(head, null)
  }

}

File: src/main/scala/kim/dencka/java/easy/collection/ReverseLinkedList.java
package kim.dencka.java.easy.collection;

import kim.dencka.java.payloads.ListNode;

class ReverseLinkedList {
    public static void main(String[] args) {
        ListNode input = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
        printAll(reverseList(input));
    }

    public static ListNode reverseList(ListNode head) {
        // Iterative
        ListNode res = null, list = null;
        while (head != null) {
            res = new ListNode(head.val, list);
            list = res;
            head = head.next;
        }
        return res;

//        return reverse(head, null);
    }
    // Recursive
    public static ListNode reverse(ListNode list, ListNode acc) {
        if (list == null) return acc;
        else return reverse(list.next, new ListNode(list.val, acc));
    }

    private static void printAll(ListNode list) {
        while (list != null) {
            System.out.print(list.val + " ");
            list = list.next;
        }
        System.out.println();
    }
}
